### DISCUSSION TITLE
AES-GCM-256 has 12B Nonce, not 8

### DISCUSSION THREAD

**Discussion by sempervictus** [minio/minio#14068] [Q&A] [answered]
After my [request](https://github.com/minio/minio/issues/14032#issuecomment-1006072055) to use industry-standard IAM for user/group CRUD was denied, i went about implementing the routines on my own in Rust. While working off of the [mdadmin encrypt source file](https://github.com/minio/madmin-go/blob/main/encrypt.go) i ran across this layout for the buffer:
```go
//    salt | AEAD ID | nonce | encrypted data
//     32      1         8      ~ len(data)
```
indicating that `nonce := sioutil.MustRandom(stream.NonceSize())` would always be 8B. [AES256-GCM uses a 12B nonce](https://soatok.blog/2020/05/13/why-aes-gcm-sucks/#:~:text=Although%20the%20AES%20block%20size,to%20actually%20encrypt%2Fdecrypt%20messages.), which i can confirm through the RustCrypto AEADs library's unwillingness to coax a relevant `Nonce` buffer from an 8B input.
Is the MinIO admin user info buffer really variable at 32B of salt, 1 of AEAD, and either 8 or 12 of nonce before the buffer, or is sioutil actually allowing the use of an under-sized nonce in the golang world? The comment in
```go
	var (
		salt  [32]byte
		id    [1]byte
		nonce [8]byte // This depends on the AEAD but both used ciphers have the same nonce length.
	)
```
strongly indicates the latter option which would be a cryptographic mis-implementation/error. As does the read call:
```go
	if _, err := io.ReadFull(data, nonce[:]); err != nil {
		return nil, err
	}
```
This is _potentially_ a security issue (i'm not crying CVE here, but someone might if i'm correct), and when used in FIPS environments is probably not meeting the FIPS spec if the cipher isn't primed correctly with a full nonce (again, not even sure how that's possible but i'm not digging into golang's crypto underbelly without a team of orderlies and a padded room ready to go).

The rust version of `sio` looks unmaintained, missing critical primitives, and basically looks to have been someone's weekend project to poke at the concept - so i can't use that for parity, nor would it be a good idea if it actually doesn't meet the cipher spec.

Ping @harshavardhana
I'm not a formal cryptographer, but i do work in commsec and am often asked to question the designs of people smarter than me for a living; so figured it's better to raise the issue than assume everything's a-OK without getting the green light from MinIO people. That SIO streaming implementation looks pretty janky (and abandoned) on the Rust side; so if their Go version is written the same way then it might warrant a re-think.
From a security architect's perspective, i'm not sure why MinIO is doing this at all: commsec (specifically the secrecy and veracity attributes) is achieved through transport encryption on production systems (TLS, NoiseSocket, IPSEC, etc). The encryption of this one buffer doesn't really help anything if the attacker has transport access to read the AK/SK from the headers of the admin user creating the new new user account. There's also the fun feature of having the nonce and salt in cleartext at the start of the ciphertext buffer on the wire along with the cipher-spec for attackers to use in brute-forcing the password (the only remaining unknown). In USG and other compliance-beholden environments sticking to crypto standards, it adds the complexity of requiring a cryptographic audit every time the code changes or the underlying dependencies change. I might be missing some relevant business logic here, but from a pure infosec perspective this appears to be somewhat redundant or by itself incomplete as a secrecy measure while potentially not adhering to the cryptographic standards it attempts to use. Given that MinIO is used so widely, what's the external entity doing the crypto audits, how often do they occur, and where can users/partners/customers evaluate the outputs?

Lastly, if i come up with something usable as a MinIO IAM-replacement in Rust for user/group CRUD, would the MinIO team want to take ownership of the library for maintenance purposes? I head-up infosec outfits and high-end consultancies with very broad-spectrum efforts, so its a bit difficult to maintain long-running things in subject domains of which i am only sometimes an inhabitant (two weeks ago i was writing various log4shell code for Metasploit, before that OpenStack in Python and before that HL7 medical data in Rust... ad nauseum).

**Comment by sempervictus** [comment]
I've now modified my code to use `chacha20poly1305` as well, and same thing - 12B (96b) [per the cipher spec RFC](https://datatracker.ietf.org/doc/html/rfc8103)

Here's the decryption function i'm using to test this, and it's blowing up on trying to allocate the non-GCM nonce for `chacha20poly1305` because the nonce buffer is under-sized by 4B:
```rust
pub async fn mio_encrypt(password: String, data: Vec<u8>, gcm: bool) -> Result<Vec<u8>, Box<dyn Error>> {
    // Crypto params
    let salt = rand::thread_rng().gen::<[u8; 32]>();
    let argon = argon2::hash_raw(password.as_bytes(), &salt, &ARGON_CONFIG).unwrap();
    let rand12 = rand::thread_rng().gen::<[u8; 12]>();
    let (nonce, key, cid) = if gcm {
        let nonce  = aes_gcm::Nonce::from_slice(&rand12[..]);
        let key    = aes_gcm::Key::from_slice(&argon[..]);
        (nonce, key, ARGON2ID_CHACHA20_POLY1305)
    } else {
        let nonce = chacha20poly1305::Nonce::from_slice(&rand12[..8]); // THIS IS WHERE WE CRASH
        let key   = chacha20poly1305::Key::from_slice(&argon[..]);
        (nonce, key, ARGON2ID_AES_GCM)
    };
    // Cleartext payload
    let payload = Payload {
        msg: &data,
        aad: b""
    };
    // Encryption with selected cipher
    let mut ciphertext = if gcm {
        let cipher = aes_gcm::Aes256Gcm::new(key);
        cipher.encrypt(nonce, payload).unwrap()
    } else {
        let cipher = chacha20poly1305::ChaCha20Poly1305::new(key);
        cipher.encrypt(nonce, payload).unwrap()   
    };
    // let (ct, tag) = ciphertext.split_at(ciphertext.len() - 16);
    // Build final buffer
    let mut encbuf = Vec::<u8>::new();
    encbuf.append(&mut salt.clone().to_vec());
    encbuf.push(cid);
    encbuf.append(&mut nonce.to_vec());
    encbuf.append(&mut ciphertext);
    Ok(encbuf)

}

pub async fn mio_decrypt(password: String, data: Vec<u8>) -> Result<Vec<u8>, Box<dyn Error>> {
    // Parse buffer elements
    let salt = &data[0..32].to_vec();
    let cid = &data[32];
    let gcm: bool = cid == &ARGON2ID_CHACHA20_POLY1305;
    let (nonce, ptr) = if gcm { 
        (aes_gcm::Nonce::from_slice(&data[33..45]), 45) // GCM256 nonce is 12B 
    } else {
        (chacha20poly1305::Nonce::from_slice(&data[33..41]), 41)
    };
    let ctext = &data[ptr..].to_vec();
    // Build key
    let argon = argon2::hash_raw(password.as_bytes(), salt, &ARGON_CONFIG).unwrap();
    let key = aes_gcm::Key::from_slice(&argon[..]);
    // Ciphertext payload
    let payload = Payload {
        msg: &ctext,
        aad: b"",
    };
    // Decryption with selected cipher
    let plaintext = if gcm {
        let cipher = aes_gcm::Aes256Gcm::new(key);
        cipher.decrypt(nonce, payload).unwrap()
    } else {
        let cipher = chacha20poly1305::ChaCha20Poly1305::new(key);
        cipher.decrypt(nonce, payload).unwrap()
    };
    Ok(plaintext)
}
```
and this is it blowing up on the `Nonce` input vector being 8B:
```rust
thread 'tests::verify_chacha_crypt' panicked at 'assertion failed: `(left == right)`
  left: `8`,
 right: `12`', /user/.cargo/registry/src/github.com-1ecc6299db9ec823/generic-array-0.14.5/src/lib.rs:559:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:138:5
   4: <&generic_array::GenericArray<T,N> as core::convert::From<&[T]>>::from
             at /user/.cargo/registry/src/github.com-1ecc6299db9ec823/generic-array-0.14.5/src/lib.rs:559:9
   5: <T as core::convert::Into<U>>::into
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/convert/mod.rs:542:9
   6: generic_array::GenericArray<T,N>::from_slice
             at /user/.cargo/registry/src/github.com-1ecc6299db9ec823/generic-array-0.14.5/src/lib.rs:541:9
   7: minio_adm_rs::mio_encrypt::{{closure}}
             at ./src/lib.rs:39:21
```

**Comment by klauspost** [comment]
Added @aead who is the authority on crypto.

The code would not work if what you write is true:
```
func TestEncryptData1(t *testing.T) {
	salt := sioutil.MustRandom(32)
	var stream *sio.Stream

	password := "1234"
	key := pbkdf2.Key([]byte(password), salt, pbkdf2Cost, 32, sha256.New)
	stream, _ = sio.AES_256_GCM.Stream(key)
	t.Log(stream.NonceSize())

	key = argon2.IDKey([]byte(password), salt, argon2idTime, argon2idMemory, argon2idThreads, 32)
	stream, _ = sio.AES_256_GCM.Stream(key)
	t.Log(stream.NonceSize())

	stream, _ = sio.ChaCha20Poly1305.Stream(key)
	t.Log(stream.NonceSize())
}
```

Outputs:
```
=== RUN   TestEncryptData1
    encrypt_test.go:111: 8
    encrypt_test.go:115: 8
    encrypt_test.go:118: 8
```

The full nonce is written. You can simply switch on the ID if you need to read back different nonce sizes.

**Comment by sempervictus** [comment]
Thanks for pinging back - so that `stream.NonceSize()` call _should_ say 12, but in the MinIO buffer read it specifies 8.

**Comment by klauspost** [comment]
IDK. The nonce is reported by https://github.com/secure-io/sio-go - https://github.com/secure-io/sio-go/blob/master/sio.go#L147

**Comment by sempervictus** [comment]
Per the spec those should be 12B for both chacha and gcm256, but the minio code doesn't care about that size in its reading of the buffer - far as my paltry go skills tell me, thats a static read of 8B. So it could be that sio is padding up 4B of zeroes on that short-read, some other error-handling magic is occurring, the sio lib doesn't support the cipher spec correctly, or that the meaning of `nonce` in this context is different from the internal cipher state requirements described in the algo's RFCs.
Apologies if this becomes time-sensitive, but i'm likely AFK for the next 6-8h or so while on the move - will check email for notifications and grab a netbook just in case.

**Comment by harshavardhana** [comment]
It is not that CRUD style is something that we overlooked or that we didn't want to implement. We looked at IAM APIs seriously and found them to be either completely over-complicated or don't meet our goals. If there is a real requirement that comes in the future we may implement AWS IAM compliant APIs but right now it's not our scope. 

For us, if you are an end-user you should use `mc admin` commands that's all and that's where it ends - beyond that if you want to write madmin in rust is your hobby requirement we have no interest in maintaining or even promoting it. 

`mc` is designed to be automated provides a nice `--json` output which allows for you to script things around. This is all that is needed. We also have a nice Console UI that provides more than what CLI provides as well. 

Beyond this engineering this yourself is a waste of time and I don't even know why you want to spend time doing these things. Your time is well spent in architecting MinIO deployments.

> IDK. The nonce is reported by https://github.com/secure-io/sio-go - https://github.com/secure-io/sio-go/blob/master/sio.go#L147

SIO is a custom format explained here https://github.com/minio/sio/blob/master/DARE.md#3-package-format you should read that first @sempervictus

**Comment by sempervictus** [comment]
@harshavardhana: thanks for jumping in.
I understand that the paradigm currently in play for "users" is the `mc` CLI, but unfortunately that is not an option when the "user" is a single static binary written in another language and carrying its own embedded libc (so linking against Go libs which link against the OS' libc would likely be a mess). The effort is not a spare-time toy, nor a rewrite of `mc` - this is for production use, and required to implement MinIO instead of a keystone-integrated Ceph with S3/Swift via RGW in the context of dynamic user/group CRUD based on the input data being generated for upload (users have to be created and policies assigned just-in-time for the relevant subjects to exist for policy to be created for data being sent to S3). Siraj has more details on the what/why of this thing.

In terms of the DARE piece: 
1. Thank you for pointing me to that doc - wouldn't have figured out that this was storage crypto for a bit there since i'm just trying to deal with a small subset of transport concern.
2. if i'm reading this correctly then the design does in fact assume that exposure of the nonce will occur and relies on a common one across blocks stored on disk - makes sense for that use-case, but looks "quite strange" when dealing with a single small message transmitted to the MinIO server.
3. Given that this is a mechanism designed for on-disk storage, and that the request is traversing the wire inside a TLS tunnel which is designed for transport-level secrecy, why is the `setUser` mechanism using this at all? Trying to understand the intent/mindset behind the design.
4. The DARE document seems to describe a slightly different encryption/decryption routine than what is actually implemented in [madmin-go](https://github.com/minio/madmin-go/blob/main/encrypt.go#L102) as [the header/message block structure described by the DARE doc](https://github.com/minio/sio/blob/master/DARE.md#3-package-format) does not conform to the salt+id+nonce structure used by the `madmin-go` code. Is it actually the DARE format and i'm misreading something, or are they distinct?

Lastly, skimming through the SIO code you linked, it does in fact truncate the nonce by 4B which seems to have something to do with the chaining mechanism for the stream elements - will need to dig into this further to try and mock-up a single block of that stream writer's output in Rust. Without further digging and review i can't say whether or not the core concern around using an 8B nonce is valid, so looks like i've got some Go to read. Much appreciate the pointers sir.

**Comment by harshavardhana** [comment]
> I understand that the paradigm currently in play for "users" is the `mc` CLI, but unfortunately that is not an option when the "user" is a single static binary written in another language and carrying its own embedded libc (so linking against Go libs which link against the OS' libc would likely be a mess). The effort is not a spare-time toy, nor a rewrite of `mc` - this is for production use, and required to implement MinIO instead of a keystone-integrated Ceph with S3/Swift via RGW in the context of dynamic user/group CRUD based on the input data being generated for upload (users have to be created and policies assigned just-in-time for the relevant subjects to exist for policy to be created for data being sent to S3). Siraj has more details on the what/why of this thing.

The market leader is AWS and we kept our APIs very close to how AWS IAM would implement things - https://docs.aws.amazon.com/IAM/latest/APIReference/API_AddUserToGroup.html - is sort of similar to https://github.com/minio/minio/blob/dabb0581678555cfbef38f3a70395d918822a9d8/cmd/iam.go#L1166

So in the future things to make it easy for us to implement IAM if needed. It is fairly easy for us to give what AWS IAM equivalent but that is not yet planned.  Unless there is a real production requirement. 

I will check with Sriraj.

**Comment by sempervictus** [comment]
I by no means want to ask you guys to shift gears and throw an IAM implementation ahead of anything already on the docket - its a brutal task and if it does get executed, is not currently at a state of priority to preempt anything else. Its less work to update sio-rs or extract the relevant pieces from it to be able to execute the basic functions needed to deliver client functionality anyway. That said, if there's any voting pool for getting IAM on the schedule - please count me as a "yes" :smile: 
I took a look through the generic trait implementations for the old `sio-rs` lib and and confirmed that the nonce really is full-width internally when enciphered but the[ last 4B of the 12 needed are taken from the sequence ID of the block being encrypteed](https://github.com/secure-io/sio-rs/blob/master/src/aead.rs#L120). So with that in-mind, if the underlying SIO code for Go is doing the same thing, then we technically do have an exposure concern since the seq ID and nonce bytes are visible in the block at known positions but the risk appears to be accepted in the design. Given that the underlying nonce structure **is** full width though, i think that the core cryptographic concern is addressed and that the implementation concerns are up to the users to qualify as safe or not for their environment.

**Comment by sempervictus** [comment]
So here's the prototype of a single message handler which hard-codes the sequence ID to zero:
```rust
pub async fn mio_encrypt(password: String, data: Vec<u8>, gcm: bool) -> Result<Vec<u8>, Box<dyn Error>> {
    // Crypto params
    let salt = rand::thread_rng().gen::<[u8; 32]>();
    let argon = argon2::hash_raw(password.as_bytes(), &salt, &ARGON_CONFIG).unwrap();
    let mut seqid = Vec::from([0u8,0,0,0]); // First message in "sequence"
    let mut nbytes = rand::thread_rng().gen::<[u8; 8]>().to_vec();
    let mut noncebytes = nbytes.clone();
    noncebytes.append(&mut seqid);
    let (nonce, key, cid) = if gcm {
        let nonce  = aes_gcm::Nonce::from_slice(&noncebytes[..]);
        let key    = aes_gcm::Key::from_slice(&argon[..]);
        (nonce, key, ARGON2ID_CHACHA20_POLY1305)
    } else {
        let nonce = chacha20poly1305::Nonce::from_slice(&noncebytes[..]);
        let key   = chacha20poly1305::Key::from_slice(&argon[..]);
        (nonce, key, ARGON2ID_AES_GCM)
    };
    // Cleartext payload
    let payload = Payload {
        msg: &data,
        aad: b""
    };
    // Encryption with selected cipher
    let mut ciphertext = if gcm {
        let cipher = aes_gcm::Aes256Gcm::new(key);
        cipher.encrypt(nonce, payload).unwrap()
    } else {
        let cipher = chacha20poly1305::ChaCha20Poly1305::new(key);
        cipher.encrypt(nonce, payload).unwrap()   
    };
    // Build final buffer
    let mut encbuf = Vec::<u8>::new();
    encbuf.append(&mut salt.clone().to_vec());
    encbuf.push(cid);
    encbuf.append(&mut nbytes);
    encbuf.append(&mut ciphertext);
    Ok(encbuf)

}

pub async fn mio_decrypt(password: String, data: Vec<u8>) -> Result<Vec<u8>, Box<dyn Error>> {
    let mut seqid = Vec::from([0u8,0,0,0]); // First message in "sequence"
    // Parse buffer elements
    let salt = &data[0..32].to_vec();
    let cid = &data[32]; // unused for now, selector for ciphers
    let gcm: bool = cid == &ARGON2ID_CHACHA20_POLY1305;
    let mut noncebytes = data[33..41].to_vec();
    noncebytes.append(&mut seqid);
    let nonce = if gcm { 
        aes_gcm::Nonce::from_slice(&noncebytes[..])
    } else {
        chacha20poly1305::Nonce::from_slice(&noncebytes[..])
    };
    let ctext = &data[41..].to_vec();
    // Build key
    let argon = argon2::hash_raw(password.as_bytes(), salt, &ARGON_CONFIG).unwrap();
    // Ciphertext payload
    let payload = Payload {
        msg: &ctext,
        aad: b"",
    };
    // Decryption with selected cipher
    let plaintext = if gcm {
        let key = aes_gcm::Key::from_slice(&argon[..]);
        let cipher = aes_gcm::Aes256Gcm::new(key);
        cipher.decrypt(nonce, payload).unwrap()
    } else {
        let key = chacha20poly1305::Key::from_slice(&argon[..]);
        let cipher = chacha20poly1305::ChaCha20Poly1305::new(key);
        cipher.decrypt(nonce, payload).unwrap()
    };
    Ok(plaintext)
}
```
which is passing these tests:
```rust
    #[test]
    fn verify_gcm_crypt() {
        let password  = "secret".to_owned();
        let plaintext = "this is a string".as_bytes().to_vec();
        let ciphertext = executor::block_on(mio_encrypt(password.clone(), plaintext.clone(), true)).unwrap();
        let decrypted  = executor::block_on(mio_decrypt(password, ciphertext)).unwrap();
        assert_eq!(plaintext,decrypted);
    }
    #[test]
    fn verify_chacha_crypt() {
        let password  = "secret".to_owned();
        let plaintext = "this is a string".as_bytes().to_vec();
        let ciphertext = executor::block_on(mio_encrypt(password.clone(), plaintext.clone(), false)).unwrap();
        let decrypted  = executor::block_on(mio_decrypt(password, ciphertext)).unwrap();
        assert_eq!(plaintext,decrypted);
    }
```
Final version will probably just take a sequence ID as a `u8` but this should be enough to stub a test against MinIO once i figure out which request scaffold to use (sorta hoping to wrap this in Rusoto code).

**Comment by sempervictus** [comment, marked as answer]
To follow up on this for anyone who tries to do it on their own - the `sio` code used in the Go admin repo is from the secure-io repo, not the minio sio repo. The underlying incremented 4B nonce-tail mechanism and actual buffer sizes for chained encryption are one and the same between the `sio-rs` and Go versions despite the Rust version being somewhat out of date. I was able to finally get this to work in the `minio-rs` codebase in [my repository](https://github.com/sempervictus/minio-rs/tree/feature/sio_user_management) if anyone's interested or needs to dynamically create users like we do from Rust.
